misk/src/main/kotlin/misk/jdbc/Cluster.kt

```diff
@@ -0,0 +1,23 @@
+package misk.jdbc
+
+/**
+ * [Cluster] provides a consistent abstraction for
+ * interacting with various database-related resources [T]
+ * based on their role in the underlying database topology.
+ *
+ * Database clusters are often composed of a Read/Write master
+ * instance and a set of Read-Only replica instances.
+ *
+ * Misk defers management of these replicas to the underlying
+ * persistence layer and requires a single addressable endpoint.
+ *
+ * Examples include a Read-Only [DataSource] or a Read/Write
+ * [SessionFactory] in Hibernate.
+ *
+ * N.B. If a Read-Only resource is not provided Misk will
+ * default to the Read/Write resource.
+ */
+interface Cluster<T> {
+  val writer: T
+  val reader: T
```

Comment:
With Vitess we would not need to have separate readers and writers. You can have a single connection and change whether you're reading from replicas or from the master with a `USE` statement. `USE @replica` makes your connection start doing replica reads, `USE @master` takes you back to the master.

I'm imagining we would instead have methods on the transacter that looks like this:
```
transacter.replicaRead().executeWithBackoff(s -> {
  s.createQuery(...)
  ...
})
```

(Sorry not valid Kotlin, I'm still a noob there.)

And this for master reads:
```
transacter.consistentRead().execute(...)
```

For slow OLAP style queries:
```
transacter.replicaRead().slow().execute(...)
```

And even in the future:
```
transacter.consistentReplicaRead().execute(...)
```
If Vitess implements this: http://www.proxysql.com/blog/proxysql-gtid-causal-reads

/cc @mmihic